### ISSUE TITLE
Potential Optional Chaining Bug

### ISSUE THREAD

**Issue by MercerK** [microsoft/TypeScript#41585]
**TypeScript Version:**  3.7.x-dev.201xxxxx

**Search Terms:**

Optional Chaining
Type bug

**Code**'

We found a bug within our codebase where a specific instance is returning "undefined" within production code. Upon review, we found this instance and recreated it here. Basically, TypeScript knows that "value" is associated to "string[] | undefined", however, it ignores "undefined" when ".includes" is added onto it for most instances and acts like it is "string[]".

```ts
interface Model {
    MyField: FieldModel
}

interface FieldModel {
    value: string[]
}

type ValidateFn<Model, Key extends keyof Model> = (
    value: Partial<Model[Key]>,
    attributes: Partial<Model>
) => void


const validator:ValidateFn<Model, 'MyField'> = (v, attributes) => {
    // value is 'string[] | undefined', yet, shows no error for value/includes.
    attributes.MyField?.value.includes('test')

    const MyField = attributes.MyField
    // value is 'string[] | undefined', yet, shows no error for value/includes.
    MyField?.value.includes('test')

    const value = attributes.MyField?.value
    // value is 'string[] | undefined'. This is the only one that shows an error for value/includes.
    value.includes('test')
}
```

**Expected behavior:**

Expected an error associated to "value" in each instance as value is a "string[] | undefined". 

**Actual behavior:**

An error is only returned on the last line. The previous lines are not showing an error. Thus, this would create a false positive that "undefined" is handled to the developer (unless they test or check it manually). 

**Playground Link:** 

[Playground](https://www.typescriptlang.org/play?ts=4.2.0-dev.20201112#code/JYOwLgpgTgZghgYwgAgLIHsAmEA2yDeAUMiWgJ4BiwumAXMlTRtjoQL6GGiSyIqM5MzXAWKkAbnBwBXCPQDOYKKADmAbQC67TmDIAHFADUpwTHEgUQAHmE4ANMgDSEMsggAPSCEzzkAaxd0GDQsXAA+ZABeZAAKMRJJGTlkAAU4KDBgKRtQnDVnMg0wu3jkcyVgACNpSHl6NIysnByWMMIASiiI8XRTTkIEdBBFZETTc3QoWmMccYtrWwcAclRKakEliOiY8Qdy5WrazsiIolJkAHoL0alZZGBfJcVlEHUNZAAfZGlvCBhQCCYJYOMgQMAOeQAC3QAHdfCB0G4oFBJsgYKjErILqAEDJsPIAHSlfZVGoQQmrASYAD8BMxEAJOLx5JiS1qYCW7U450GwzA5CpUTKYAqh3JBMp60wpSuNyS90ez1Umk+31+-xAgOByFB4OQUNh8MR0BRUDRGNuEGxIFx0nxRPOkpotPpjJtzPkrPZnO5pF5I3pQpJYopa2ddMtMuugYeyCeFVeKq+P2wGq1ILBEOhcLKICRpvNZvp1tt9tKrqZdpZbPJHK5bCAA)

**Related Issues:**

**Comment by DanielRosenwasser** [comment]
I think this is actually an issue with quick info. None of the intermediate property accesses should display with the type `| undefined`

**Comment by RyanCavanaugh** [comment]
It looks like we're incorrectly carrying the optional chain undefined marker through the end of the expression. The subtlety is that
```
({})?.foo?.bar
```
evaluates to `undefined` but
```
({})?.foo?.bar()
```
does *not* throw - but we're evaluating this like `foo?.bar.baz()` won't throw (it will!)

Simplified
```ts
interface Model {
    MyField?: { value: string[] }
}

function fn(attributes: Model) {
    // value is 'string[] | undefined', yet, shows no error for value/includes.
    attributes.MyField?.value.includes('test')

    const MyField = attributes.MyField
    // value is 'string[] | undefined', yet, shows no error for value/includes.
    MyField?.value.includes('test')

    const value = attributes.MyField?.value
    // value is 'string[] | undefined'. This is the only one that shows an error for value/includes.
    value.includes('test')
}
```

**Comment by rbuckton** [comment]
I concur with @DanielRosenwasser. if we consider this a Bug, then its a bug in QuickInfo, not the type checker. If you were to write `attributes.MyField?.value` on its own, the result would *correctly* be `string[] | undefined`. However, `.includes('test')` is part of the _OptionalChain_, and is therefore included in the branch where `MyField` is defined (and therefore `value` has type `string[]`).

@RyanCavanaugh your example does not illustrate a bug, it illustrates the behavior of optional chaining. Any `.`, `[]`, or `()` to the right of `?.` is part of the _OptionalChain_ and is *not* a regular property access.

```ts
// this is correct behavior. `?.value.includes('test') is the OptionalChain, and is only evaluated if `MyField` is defined.
attributes.MyField?.value.includes('test'); 

// this is also correct behavior, for the same reason above.
const MyField = attributes
MyField?.value.includes('test')

// this is *also* correct behavior. Here `?.value` is the OptionalChain.
const value = attributes.MyField?.value
// here, `.includes('test')` is *not* in an OptionalChain, so it should be an error.
value.includes('test')
```

**Comment by RyanCavanaugh** [comment]
I got myself confused on this; thanks for the correction.

I don't think there's a better way to display this in QuickInfo - we'd have to somehow differentiate "undefined as the final result" and "undefined as an intermediate optional chain result", which is at best going to be confusing.

**Comment by DanielRosenwasser** [comment]
> we'd have to somehow differentiate "undefined as the final result" and "undefined as an intermediate optional chain result", which is at best going to be confusing.

but if you look at the current use-case, it's totally broken. The intermediate variables are undefined, and the method returns something truthy. I really think this is a bug.

**Comment by DanielRosenwasser** [comment]
Also, how does C# do this?

**Comment by CyrusNajmabadi** [comment]
Ping me :)

**Comment by RyanCavanaugh** [comment]
@DanielRosenwasser please spec 😉

**Comment by DanielRosenwasser** [comment]
Okay, very loosely because this isn't how the syntax is represented, for

```
expr?.d1.[...].dn
```

For every `i < n`, the quick info on `di` is the same as the quick info if there had been no optional chain expression.

For `dn`,

* if `dn` is `m()` (part of a call expression), then the quick info of `dn` is generated as if the return type of the call expression has an `| undefined`
* if `dn` is an identifier, then the quick info is generated as if the property type had an `| undefined`.

be careful to "do the right thing" for element accesses too.

----

Apparently the way C# works is by _never_ changing the type from what it's declared as. I'm vaguely okay with this behavior if the above is too complex, but that seems different from how our quick info usually operates. I'd be surprised to see a different type between `x` and `b` in the following example.

```ts
let x = a?.b;
```

**Comment by CyrusNajmabadi** [comment]
>  I'd be surprised to see a different type between x and b in the following example.

I personally disagree.  These are two separate things.  If i have `a?.b` then if i hover on `b` i want to know what `b` actually is as that will dictate what i can do with it as i continue that expression.  For example `a?.b.x`  The things i can `dot off` of `b` will be based on `b`'s actual type, not the type of the entire `a?.b` expression.  

It's the difference between `a?.b.` vs `(a?.b).`.   Since there is no parenthesization, and since we're examining the RHS, i think it only makes sense to to state what `b` actually is.  If we didn't, it would actually be more confusing as it would imply that you could get a different set of results than what you actually get.

**Comment by marafat** [comment]
Hi everyone, 

We have ran into a similar case, this time with element access. It resulted in a runtime error due accessing an attribute on `undefined`. Here's an example to reproduce the issue:

```ts
interface Currency {
  code: string;
}

interface Account {
  owner: string;
  supportedCurrencies?: Currency[];
}

const account: Account = {
  owner: 'John Doe'
}

// This correctly gets the type "Currency | undefined"
const firstCurrency = account.supportedCurrencies?.[0];

// Correctly getting an error: "Object is possibly undefined"
const firstCurrencyCode = firstCurrency.code;

// Missing error: "Object is possibly undefined"
const firstCurrencyCodeDirect = account.supportedCurrencies?.[0].code;
//                                                              ^^^^^ - it should be only evaluated if the element access is defined
// however, in production it causes a runtime error because of accessing "code" on "undefined"
```

[Playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMIFcpQiBBPZAbwChlkEB7AEwgC5kBnMKUAcwG5iBfY40SWIhQBBBJXTgipZBQDuIaPSYsQHaQ3QAHTRSiQqGLDgTAIDAPz1D2XHgDaAXU49ilEE2SJx4eqO9hkAF4pMjkFKHoAcgApCgALEGQAEQoISO5eAHpM5AAVOOAGcl1sBDAAGwJWCDAisDiUMDxNFAAia2MCAB9kCRoYUAgqVtcKdwCBqCYO2yDPMQoJMAA6DW1dfRmTM3NluwAGJyyc1BKIMsrkarAwNk9E6ChdelaAeQAjACtzgMLkHQYDGA70ufQgAwUw1G42Qk2mmBs+FONDmcLAWzwy0oNE4xGyyAAsoUgapkI9nsg3l8fsg-gCgSCCGCIUMRm4PGiMciIElgKUAsEvItwKstDo9EMtqYLHtDljqBBOPiyCrVWr1RrNVrtRqAHr6-XIAC0tICDDii3KVGQ7xQY0uEAAbnByug4PpaTBkPUUBByhAALY4AJeMxFYDW-qDKh4nIW2RO6AAGlpiU0Tyo6DKwDGpvIcHQDDMnmQUCWwCDZKwuht5wLRZkXtDgLurWxEFaMkSrWZ0ZGvCAA).

We expected TS to be able to help us prevent this runtime error.

**Comment by RyanCavanaugh** [comment]
@marafat that's unrelated; please log a new issue. Thanks!

**Comment by Andarist** [comment]
@RyanCavanaugh I believe this one can be closed as it was fixed by https://github.com/microsoft/TypeScript/pull/53804
